talk/owt/sdk/base/win/ffmpeg_decoder_impl.cc

```diff
@@ -0,0 +1,570 @@
+#include "third_party/libyuv/include/libyuv/convert.h"
+#include "webrtc/system_wrappers/include/clock.h"
+
+#include "talk/owt/sdk/base/nativehandlebuffer.h"
+
+namespace owt {
+namespace base {
+
+namespace {
+
+constexpr std::array<AVPixelFormat, 2> kPixelFormatsSupported = {
+    AV_PIX_FMT_YUV420P, AV_PIX_FMT_YUV444P};
+constexpr size_t kYPlaneIndex = 0;
+constexpr size_t kUPlaneIndex = 1;
+constexpr size_t kVPlaneIndex = 2;
+
+struct ScopedPtrAVFreePacket {
+  void operator()(AVPacket* packet) { av_packet_free(&packet); }
+};
+typedef std::unique_ptr<AVPacket, ScopedPtrAVFreePacket> ScopedAVPacket;
+
+ScopedAVPacket MakeScopedAVPacket() {
+  ScopedAVPacket packet(av_packet_alloc());
+  return packet;
+}
+
+}  // namespace
+
+int FFMpegDecoderImpl::AVGetBuffer2(AVCodecContext* context,
+                                    AVFrame* av_frame,
+                                    int flags) {
+  // Set in `Configure`.
+  FFMpegDecoderImpl* decoder = static_cast<FFMpegDecoderImpl*>(context->opaque);
+  // DCHECK values set in `Configure`.
+  RTC_DCHECK(decoder);
+  // Necessary capability to be allowed to provide our own buffers.
+  RTC_DCHECK(context->codec->capabilities | AV_CODEC_CAP_DR1);
+
+  auto pixelFormatSupported = std::find_if(
+      kPixelFormatsSupported.begin(), kPixelFormatsSupported.end(),
+      [context](AVPixelFormat format) { return context->pix_fmt == format; });
+
+  RTC_CHECK(pixelFormatSupported != kPixelFormatsSupported.end());
+
+  // `av_frame->width` and `av_frame->height` are set by FFmpeg. These are the
+  // actual image's dimensions and may be different from `context->width` and
+  // `context->coded_width` due to reordering.
+  int width = av_frame->width;
+  int height = av_frame->height;
+  // See `lowres`, if used the decoder scales the image by 1/2^(lowres). This
+  // has implications on which resolutions are valid, but we don't use it.
+  RTC_CHECK_EQ(context->lowres, 0);
+  // Adjust the `width` and `height` to values acceptable by the decoder.
+  // Without this, FFmpeg may overflow the buffer. If modified, `width` and/or
+  // `height` are larger than the actual image and the image has to be cropped
+  // (top-left corner) after decoding to avoid visible borders to the right and
+  // bottom of the actual image.
+  avcodec_align_dimensions(context, &width, &height);
+
+  RTC_CHECK_GE(width, 0);
+  RTC_CHECK_GE(height, 0);
+  int ret = av_image_check_size(static_cast<unsigned int>(width),
+                                static_cast<unsigned int>(height), 0, nullptr);
+  if (ret < 0) {
+    RTC_LOG(LS_ERROR) << "Invalid picture size " << width << "x" << height;
+    decoder->ReportError();
+    return ret;
+  }
+
+  // The video frame is stored in `frame_buffer`. `av_frame` is FFmpeg's version
+  // of a video frame and will be set up to reference `frame_buffer`'s data.
+  rtc::scoped_refptr<webrtc::PlanarYuvBuffer> frame_buffer;
+  rtc::scoped_refptr<webrtc::I444Buffer> i444_buffer;
```

Comment:
No need to decorate i444_buffer and i420 buffer here? They are only used in switch case statement.

Reply:
Looks like i have to be forward declared. Otherwise compiler complains about i420_buffer/i444_buffer might skip initialization across switch labels.

Reply:
Okay. It passed all checks, ready to merge.